File: lib/rgfa/line.rb
require "set"

#
# Generic representation of a record of a RGFA file.
#
# @!macro [new] rgfa_line
#   @note
#     This class is usually not meant to be directly initialized by the user;
#     initialize instead one of its child classes, which define the concrete
#     different record types.
#
class RGFA::Line; end

# submodules of RGFA::Line::Common define methods which are included
# in line or in its subclasses
RGFA::Line::Common = Module.new
require_relative "line/common/init"
require_relative "line/common/dynamic_fields"
require_relative "line/common/writer"
require_relative "line/common/version_conversion"
require_relative "line/common/field_datatype"
require_relative "line/common/field_data"
require_relative "line/common/equivalence"
require_relative "line/common/cloning"
require_relative "line/common/connection"
require_relative "line/common/virtual_to_real"
require_relative "line/common/update_references"
require_relative "line/common/disconnection"
require_relative "line/common/validate"

class RGFA::Line
  include RGFA::Line::Common::Init
  include RGFA::Line::Common::DynamicFields
  include RGFA::Line::Common::Writer
  include RGFA::Line::Common::VersionConversion
  include RGFA::Line::Common::FieldDatatype
  include RGFA::Line::Common::FieldData
  include RGFA::Line::Common::Equivalence
  include RGFA::Line::Common::Cloning
  include RGFA::Line::Common::Connection
  include RGFA::Line::Common::VirtualToReal
  include RGFA::Line::Common::UpdateReferences
  include RGFA::Line::Common::Disconnection
  include RGFA::Line::Common::Validate
end

#
# Require the child classes
#
require_relative "line/header.rb"
require_relative "line/segment.rb"
require_relative "line/comment.rb"
require_relative "line/custom_record.rb"
require_relative "line/gap.rb"
require_relative "line/fragment.rb"
require_relative "line/edge.rb"
require_relative "line/group.rb"
require_relative "line/unknown.rb"

File: gfapy/line/line.py
from .common.construction import Construction
from .common.dynamic_fields import DynamicFields
from .common.writer import Writer
from .common.version_conversion import VersionConversion
from .common.field_datatype import FieldDatatype
from .common.field_data import FieldData
from .common.equivalence import Equivalence
from .common.cloning import Cloning
from .common.connection import Connection
from .common.virtual_to_real import VirtualToReal
from .common.update_references import UpdateReferences
from .common.disconnection import Disconnection
from .common.validate import Validate
from .common.default_record_definition import DefaultRecordDefinition

class Line(Construction, DynamicFields, Writer, VersionConversion,
           FieldDatatype, FieldData, Equivalence, Cloning, Connection,
           VirtualToReal, UpdateReferences, Disconnection, Validate,
           DefaultRecordDefinition):
  """
  A line of a GFA file.

  Parameters:
    data (str, list of str) : the content of a line in a GFA file, either as
      a string, or as a list derived from tab-splitting the line string
    vlevel (int) : an integer from 0 to 3, which specifies the validation level;
      if 0, no validation is performed (the user can still validate manually if
      needed); if 1 (the default), validation is performed when the line
      is constructed, or, for some fields, when the value is accessed
      for the first time; if 2, the validation is performed also when converting
      the content of a field to string; if 3, also each time the value
      of a field is read or written
    version (str) : one of 'gfa1' and 'gfa2'; the GFA version; if not specified,
      then the version is guessed from the record type and syntax, or set
      to 'generic'
    dialect (str) : one of 'rgfa' and 'standard'; the GFA dialect; if not
      specified then the dialect is set to 'standard'; 'rgfa' is only valid
      when version is 'gfa1'

  Notes:
    The private interface to the Line constructor also allows to pass a
    dictionary instead of a list for data. Furthermore the private parameter
    virtual allows to create virtual line instances, which are useful during
    parsing.

  Raises:
    gfapy.error.FormatError: If the line contains a wrong number of positional
      fields, or if the content of a field has a wrong format.
    gfapy.error.NotUniqueError: If a tag name is used more than once.
    gfapy.error.TypeError: If the value of a predefined tag does not
      respect the datatype specified in the tag.

  Returns:
    an instance of a subclass of gfapy.line.Line
  """

  SEPARATOR = "\t"
  """Separator in the string representation of GFA lines"""

